File: bot-server/crypto.py
"""
Шифрование/расшифровка токенов ботов — AES-256-GCM
"""
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _get_key() -> bytes:
    hex_key = os.environ["TOKEN_ENCRYPTION_KEY"]
    return bytes.fromhex(hex_key)


def encrypt_token(plaintext: str) -> str:
    """Шифруем токен бота перед сохранением в БД."""
    key = _get_key()
    aesgcm = AESGCM(key)
    nonce = os.urandom(12)  # 96-bit nonce
    ciphertext = aesgcm.encrypt(nonce, plaintext.encode(), None)
    # Возвращаем base64(nonce + ciphertext)
    return base64.b64encode(nonce + ciphertext).decode()


def decrypt_token(encrypted: str) -> str:
    """Расшифровываем токен бота из БД."""
    key = _get_key()
    aesgcm = AESGCM(key)
    data = base64.b64decode(encrypted)
    nonce = data[:12]
    ciphertext = data[12:]
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    return plaintext.decode()

File: supabase/functions/_shared/crypto.ts
/**
 * AES-256-GCM шифрование для хранения токенов ботов.
 * Тот же алгоритм что в bot-server/crypto.py — ключи взаимозаменяемы.
 */

const KEY_HEX = Deno.env.get("TOKEN_ENCRYPTION_KEY") ?? "";

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

async function getKey(): Promise<CryptoKey> {
  return crypto.subtle.importKey(
    "raw",
    hexToBytes(KEY_HEX),
    { name: "AES-GCM" },
    false,
    ["encrypt", "decrypt"],
  );
}

/** Зашифровать токен бота → base64(nonce + ciphertext) */
export async function encryptToken(plaintext: string): Promise<string> {
  const key = await getKey();
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const encoded = new TextEncoder().encode(plaintext);
  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    encoded,
  );
  const combined = new Uint8Array(12 + ciphertext.byteLength);
  combined.set(nonce, 0);
  combined.set(new Uint8Array(ciphertext), 12);
  return btoa(String.fromCharCode(...combined));
}

/** Расшифровать base64(nonce + ciphertext) → токен бота */
export async function decryptToken(b64: string): Promise<string> {
  const key = await getKey();
  const combined = Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
  const nonce = combined.slice(0, 12);
  const ciphertext = combined.slice(12);
  const plaintext = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    ciphertext,
  );
  return new TextDecoder().decode(plaintext);
}
